050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,717 @@
+   Exactly one of `STRIMZI_NAMESPACE` or `STRIMZI_POLICY_YAML` must be provided.
+2. Although these are present in the BTO source code they have never actually been used.
+
+#### CO-managed case
+
+```yaml
+kind: Kafka
+spec:
+  # ...
+  topicOperator:
+
+    # Unused options will be deprecated and ignored
+    topicMetadataMaxAttempts: 
+    zookeeperSessionTimeoutSeconds:
+  
+    # Options with same schema, but which may beed to be reconfigured
+    livenessProbe: # probe values may need to be changed
+    readinessProbe: # probe values may need to be changed
+    startupProbe: # probe values may need to be changed
+    resources: # Resource requirements may differ
+    jvmOptions: # JVM options may need to be changed
+    logging:  # Logger names will need to be changed
+
+    # Unchanged options
+    image: 
+    watchedNamespace: # becomes optional
+    reconciliationIntervalSeconds:
+
+    # New options
+    policy: # optional namespace policy
+```
+
+### CR API
+
+The CR API is structurally unchanged except:
+
+* `spec.managed` (described above), optional, is added
+
+Furthermore within the same structure:
+
+* `status.conditions` may have new values for the `reason`.
+
+### Metrics
+
+<table>
+<tr><th>Metric type	<th>Metric name				<th>Status
+<tr><td>counter		<td>strimzi.reconciliations.periodical	<th>Unchanged
+<tr><td>counter		<td>strimzi.reconciliations		<th>Unchanged
+<tr><td>counter		<td>strimzi.reconciliations.failed	<th>Unchanged
+<tr><td>counter		<td>strimzi.reconciliations.successful	<th>Unchanged
+<tr><td>gauge		<td>strimzi.resources			<th>Unchanged
+<tr><td>timer		<td>strimzi.reconciliations.duration	<th>Unchanged
+<tr><td>gauge		<td>strimzi.reconciliations.paused	<th>Unchanged
+<tr><td>counter		<td>strimzi.reconciliations.locked	<th>Unchanged
+<tr><td>gauge		<td>strimzi.resource.state		<th>Unchanged
+
+### Emitted Kube `Events`
+
+Let's consider all the cases in which the BTO emits `Event` resources to the Kube apiserver:
+
+* For errors when accessing the topic store. 
+  For UTO there is no topic store, so this cannot happen.
+* For errors when creating partitions. 
+  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
+  For other errors application logging and metrics are considered sufficient.
+* For errors when changing the configuration of topics in Kafka
+  For UTO ApiExceptions from the Admin client will be reported via `Ready` condition.
+  For other errors application logging and metrics are considered sufficient.
+* For errors when processing a ZooKeeper znode watch.
+  UTO doesn't use znode watches, so this cannot happen.
+* On attempts to change a `KafkaTopic`'s `spec.topicName`.
+  For UTO this will be reported via the status only.
+* On detection of conflicting changes between Kube and Kafka
+  For UTO such conflicts cannot happen.
+* On attempts to decreate the number of partitions.
+  For UTO this will be reported via the status only.
+
+Thus, the UTO will emit no events.
```

Comment:
I find that Kubernetes events provide a different dimension to native kube applications observability than logging, metrics, and status conditions. 
1. **Compared to logs & Metrics** : UTO logs and metrics are meant to be infrastructure status targeting the cluster administrators. In contrast, namespaced Kube events target the CR authors who might not have access to the CO logs and metrics. This becomes increasingly important for operators that provide some form of namespace policies or multi-tenancy-like operation mode such as the UTO. 
2. **Compared to status** Kube events provide more of a historical view of state change or error progression. For E.g. a current `Ready` resource with 0 events is different from a `Ready` resource with associated Kube events that show e.g. conflict -> ready -> conflict. Especially for a case when another resource with a conflicting name is created that ends up in **both** resources becoming not ready.

Reply:
@devguyio that's true, but I don't see any of those existing BTO events really requiring an `Event`. The one example you give (`ResourceConflict`) can only happen as a result of someone creating a conflicting `KafkaTopic` in the same namespace as an existing one. Presumably they're looking for the `Ready` in that case and will resolve the problem more or less immediately. Those resources being in conflict doesn't have any impact on how the topic is being used by Kafka applications. So I just don't seem much value in the historical state change in that case.

I think there might be a role for `Events` in the future. For example if/when we add support for RF change then when data movement is required it can take a long time. There seems a stronger case for using events to notify about those state changes (perhaps in addition to a 2nd condition with `type: ReplicasReady` perhaps), because we don't expect anyone to waiting on the outcome. But I don't want to say that in _this_ proposal, because I don't want to bite off RF change at the same time as a whole new operator.

Reply:
Yeah this makes sense to me.